packages/agents/src/agents/Chameleon/ContextualizedChat.ts

```diff
@@ -0,0 +1,294 @@
+import {
+  Chat,
+  ChatLogType,
+  ChatMessage,
+  MessageChunker,
+  LocalVectorDB,
+  LocalCollection
+} from "@evo-ninja/agent-utils";
+import { v4 as uuid } from "uuid";
+
+type MsgPtr = number;
+
+interface ChunkMetadata {
+  msgPtr: MsgPtr;
+}
+
+type ChunkMetadataPtr = number;
+
+interface DocumentMetadata {
+  chunkMetadataPtr: ChunkMetadataPtr;
+  tokens: number;
+}
+
+export class ContextualizedChat {
+
+  private _chunkMetadata: Record<ChatLogType, ChunkMetadata[]> = {
+    "persistent": [],
+    "temporary": []
+  };
+
+  private _collections: Record<ChatLogType, LocalCollection<DocumentMetadata>>;
+
+  constructor(
+    private _chat: Chat,
+    private _chunker: MessageChunker,
+    db: LocalVectorDB
+  ) {
+    this._collections = {
+      persistent: db.addCollection<DocumentMetadata>(uuid()),
+      temporary: db.addCollection<DocumentMetadata>(uuid())
+    };
+  }
+
+  get rawChat(): Chat {
+    return this._chat;
+  }
+
+  async contextualize(context: string, tokenLimits: Record<ChatLogType, number>): Promise<Chat> {
+    // Ensure all new messages have been processed
+    await this._processNewMessages();
+
+    // Aggregate all "persistent" messages
+    const persistentSmallMsgs = this._aggregateSmallMsgs(
+      "persistent",
+      tokenLimits["persistent"]
+    );
+    const persistentLargeMsgs = await this._contextualizeMsgs(
+      context,
+      "persistent",
+      tokenLimits["persistent"] - persistentSmallMsgs.tokens
+    );
+
+    // join persistent msgs
+    const persistentMsgs = {
+      msgs: [...persistentSmallMsgs.msgs, ...persistentLargeMsgs.msgs],
```

Comment:
NIT: I'd sort before saving information that's not usable in its current state in an object, just to later transform it